In [161]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [162]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [163]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [164]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [165]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaler = StandardScaler()

scaled_data = scaler.fit_transform(df_market_data)

In [166]:
# Create a DataFrame with the scaled data
df_scaled = pd.DataFrame(
    scaled_data, 
    columns=df_market_data.columns  
)

# Copy the crypto names from the original DataFrame
df_scaled["coin_id"] = df_market_data.index 

# Set the coin_id column as index
df_scaled.set_index("coin_id", inplace=True)

# Display the scaled DataFrame
df_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [167]:
# Create a list with the number of k-values from 1 to 11
k_values = range(1, 12)

In [168]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for k in range(1, 12):
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans = KMeans(n_clusters=k, random_state=0)

# 2. Fit the model to the data using `df_market_data_scaled`
kmeans.fit(df_scaled)

# 3. Append the model.inertia_ to the inertia list
inertia.append(kmeans.inertia_)
# Display the inertia values
inertia

c:\Users\abiga\anaconda3\envs\python=3.10\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


[23.890044355377952]

In [169]:

k_values = list(range(1, len(inertia) + 1))  # Generate k-values dynamically based on inertia length

# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": range(1, len(inertia) + 1), "inertia": inertia}
elbow_df = pd.DataFrame(elbow_data)

# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

# Display the DataFrame
elbow_df


k    inertia
0  1  23.890044

In [170]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.

df_scaled["Cluster"] = kmeans.labels_

scatter_plot = df_scaled.hvplot.scatter(
    x="price_change_percentage_24h",         
    y="price_change_percentage_7d",          
    by="Cluster",                            
    hover_cols=["coin_id"],                  
    title="Cryptocurrency Clusters",
    xlabel="24h Price Change (%)",
    ylabel="7d Price Change (%)",
    size=10                                  
)

# Display the plot
scatter_plot


:NdOverlay   [Cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

In [171]:
pca = PCA(n_components=2)
pca_data = pca.fit_transform(df_scaled.drop(columns=["Cluster"]))

# Create a DataFrame for the PCA results
pca_df = pd.DataFrame(
    pca_data, 
    columns=["PC1", "PC2"], 
    index=df_scaled.index
)

# Add Cluster and coin_id to the PCA DataFrame
pca_df["Cluster"] = df_scaled["Cluster"]
pca_df["coin_id"] = df_scaled.index

# Plot the scatter plot using hvplot
scatter_plot = pca_df.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="Cluster",                  # Color by cluster labels
    hover_cols=["coin_id"],        # Add coin_id to hover information
    title="Cryptocurrency Clusters (PCA)",
    xlabel="Principal Component 1",
    ylabel="Principal Component 2",
    size=10                        # Adjust marker size
)

# Display the scatter plot
scatter_plot

:NdOverlay   [Cluster]
   :Scatter   [PC1]   (PC2,coin_id)

In [172]:
elbow_plot = elbow_df.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Method for Optimal k",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    line_width=2
)
elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** k = 4
Best k is k = 4 for the original data and k=3 for the PCA data.

PCA simplified the data, leading to a different clustering structure.


---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [173]:
# Initialize the K-Means model using the best value for k
kmeans_original = KMeans(n_clusters=4, random_state=0)

In [174]:
# Fit the K-Means model using the scaled DataFrame
kmeans_original.fit(df_scaled)

# Add the cluster labels to the DataFrame
df_scaled["Cluster"] = kmeans_original.labels_

# Display the first few rows of the DataFrame with the cluster labels
df_scaled.head()


c:\Users\abiga\anaconda3\envs\python=3.10\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Cluster  
coin_id                                            
bitcoin                        -0.251637        1  
ethereum                       -0.199352        1  
tether                         -0.282061        1  
ripple                         -0.295546        1  
bitcoin-cash                   -0.270317        2

In [175]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
clusters = kmeans_original.predict(df_scaled)

# Print the resulting array of cluster values.
print(clusters)

[1 1 0 0 1 1 1 1 0 0 0 0 0 1 0 0 0 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 3 0 0 0 1
 0 0 0 0]


In [176]:
# Create a copy of the scaled DataFrame
df_scaled_copy = df_scaled.copy()
df_scaled_copy.head()



price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Cluster  
coin_id                                            
bitcoin                        -0.251637        1  
ethereum                       -0.199352        1  
tether                         -0.282061        1  
ripple                         -0.295546        1  
bitcoin-cash                   -0.270317        2

In [177]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_scaled_copy["Cluster"] = clusters

# Display the copy of the scaled DataFrame
df_scaled_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Cluster  
coin_id                                            
bitcoin                        -0.251637        1  
ethereum                       -0.199352        1  
tether                         -0.282061        0  
ripple                         -0.295546        0  
bitcoin-cash                   -0.270317        1

In [178]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
# Create a scatter plot using hvPlot
scatter_plot = df_scaled_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="Cluster",                 # Color points by cluster labels
    hover_cols=["coin_id"],       # Add cryptocurrency name to hover
    title="Cryptocurrency Clusters",
    xlabel="24h Price Change (%)",
    ylabel="7d Price Change (%)",
    size=10                       # Adjust marker size
)

# Display the scatter plot
scatter_plot


:NdOverlay   [Cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [179]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [180]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
pca_data = pca.fit_transform(df_scaled)
# down to three principal components.

pca_df = pd.DataFrame(
    pca_data,
    columns=["PC1", "PC2", "PC3"],  # Name the principal components
    index=df_scaled.index           # Retain the original index for alignment
)

# View the scaled PCA data
pca_df.head()


PC1       PC2       PC3
coin_id                                   
bitcoin      -0.469771  0.912960  0.450233
ethereum     -0.334504  0.525218  0.896458
tether       -0.488296 -0.130900 -0.516827
ripple       -0.519958 -0.177396 -0.361645
bitcoin-cash -0.586946  2.239185  2.121209

In [181]:
# Retrieve the explained variance to determine how much information

explained_variance = pca.explained_variance_ratio_

# can be attributed to each principal component.
print("Explained Variance Ratio for each Principal Component:")
for i, variance in enumerate(explained_variance, start=1):
    print(f"PC{i}: {variance:.2%}")

    # Display the total variance explained by all components
total_variance = explained_variance.sum()
print(f"\nTotal Variance Explained: {total_variance:.2%}")

Explained Variance Ratio for each Principal Component:
PC1: 36.31%
PC2: 32.84%
PC3: 19.66%

Total Variance Explained: 88.81%


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

Sum the values in explained_variance_ratio_ to calculate the total variance explained by all selected principal components.
**Answer:** Total Variance Explained: 88.81%

Explained Variance Ratio for each Principal Component:
PC1: 56.78%
PC2: 30.45%
PC3: 12.77%

Total Variance Explained: 100.00%

In [182]:
# Create a new DataFrame with the PCA data.

pca_df = pd.DataFrame(
    pca_data,  # PCA-transformed data
    columns=["PC1", "PC2", "PC3"],  # Name the principal components
    index=df_scaled.index  # Retain the original index for alignment
)
# Copy the crypto names from the original scaled DataFrame

pca_df["coin_id"] = df_scaled.index

# Set the coin_id column as index
pca_df.set_index("coin_id", inplace=True)

# Display the scaled PCA DataFrame
pca_df.head()


PC1       PC2       PC3
coin_id                                   
bitcoin      -0.469771  0.912960  0.450233
ethereum     -0.334504  0.525218  0.896458
tether       -0.488296 -0.130900 -0.516827
ripple       -0.519958 -0.177396 -0.361645
bitcoin-cash -0.586946  2.239185  2.121209

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [183]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [184]:
# Create an empty list to store the inertia values
inertia = []

df_market_data_pca = pd.DataFrame(
    pca_data, 
    columns=["PC1", "PC2", "PC3"], 
    index=df_scaled.index
)
# Create a for loop to compute the inertia with each possible value of k
for k in k_values:
    # 1. Create a KMeans model using the loop counter for the n_clusters
    kmeans = KMeans(n_clusters=k, random_state=0)
    
    # 2. Fit the model to the data using `df_market_data_pca`
    kmeans.fit(df_market_data_pca)
    
    # 3. Append the model.inertia_ to the inertia list
    inertia.append(kmeans.inertia_)

    

c:\Users\abiga\anaconda3\envs\python=3.10\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\abiga\anaconda3\envs\python=3.10\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\abiga\anaconda3\envs\python=3.10\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\abiga\anaconda3\envs\python=3.10\Lib\site-packages\sklearn\cluster\_kmeans.p

In [185]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
    "k": k_values,
    "inertia": inertia
}
# Create a DataFrame with the data to plot the Elbow curve
elbow_df = pd.DataFrame(elbow_data)

# Display the Elbow curve DataFrame
elbow_df

k     inertia
0    1  269.973377
1    2  177.984172
2    3  122.034312
3    4   55.989356
4    5   44.975033
5    6   33.403079
6    7   25.081001
7    8   18.274351
8    9   14.111068
9   10   12.649656
10  11    8.840289

In [187]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_plot = elbow_df.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Method for Optimal k",
    xlabel="Number of Clusters (k)",
    ylabel="Inertia",
    line_width=2
)

# Display the plot
elbow_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:**4

* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** no

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [188]:
# Initialize the K-Means model using the best value for k
# Initialize the K-Means model with the best k value determined earlier (e.g., k=4)
kmeans_pca = KMeans(n_clusters=4, random_state=0)


In [189]:
# Fit the K-Means model using the PCA data
kmeans_pca.fit(df_market_data_pca)

c:\Users\abiga\anaconda3\envs\python=3.10\Lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [190]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
clusters_pca = kmeans_pca.predict(df_market_data_pca)

# Print the resulting array of cluster values.
print(clusters_pca)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [191]:
# Create a copy of the scaled PCA DataFrame
df_pca_clustered = df_market_data_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
df_pca_clustered['Cluster'] = clusters_pca

# Display the copy of the scaled PCA DataFrame
df_pca_clustered.head()

PC1       PC2       PC3  Cluster
coin_id                                            
bitcoin      -0.469771  0.912960  0.450233        2
ethereum     -0.334504  0.525218  0.896458        2
tether       -0.488296 -0.130900 -0.516827        0
ripple       -0.519958 -0.177396 -0.361645        0
bitcoin-cash -0.586946  2.239185  2.121209        2

In [192]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
df_pca_clustered['coin_id'] = df_scaled.index

scatter_plot = df_pca_clustered.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="Cluster",  # Color the graph points by the cluster labels
    hover_cols=["coin_id"],  # Add the coin_id as hover information
    title="Cryptocurrency Clusters (PCA)",
    xlabel="Principal Component 1",
    ylabel="Principal Component 2",
    size=10
)

# Display the scatter plot
scatter_plot

:NdOverlay   [Cluster]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [193]:
# Composite plot to contrast the Elbow curves
# YOUR CODE HERE!

pca_scatter = df_pca_clustered.hvplot.scatter(
    x="PC1", 
    y="PC2", 
    by="Cluster", 
    hover_cols=["coin_id"], 
    title="Clusters Based on PCA Data",
    xlabel="Principal Component 1", 
    ylabel="Principal Component 2"
)

In [195]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!
scatter_plot + pca_scatter

:Layout
   .NdOverlay.I  :NdOverlay   [Cluster]
      :Scatter   [PC1]   (PC2,coin_id)
   .NdOverlay.II :NdOverlay   [Cluster]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:**   Using fewer features (PCA-reduced data) to cluster the data simplifies the clustering process and may result in clearer separation of clusters. However, it could also lead to a loss of detail and potentially less precise clustering if important features are not captured in the principal components.






